File: MD.CMS.Administration/MD.CMS.Administration.Core.Web/scripts/businessLogic/dataAccess/entities/folder.ts
/// <reference path="./base/iBaseEntity.ts" />
/// <reference path="./base/baseEntity.ts" />
/// <reference path="./../../helpers.ts" />
/// <reference path="./content.ts" />
/// <reference path="./mediaContent.ts" />
/// <reference path="./template.ts" />
/// <reference path="./profileType.ts" />
/// <reference path="./folderMetaDataField.ts" />

namespace mdBusinessLogic.dataAccess.entities {
    export class folder<T extends base.BaseEntity & base.IBaseEntity<T>> extends base.BaseEntity implements base.IBaseEntity<folder<T>> {
        public ParentId: number;
        public Name: string;
        public Description: string;
        public Parent: folder<T>;
        public Children: Array<folder<T>>;
        public Contents: Array<content>;
        public FolderPath: string;
        public MetaDataFields: Array<folderMetaDataField>;
        public MediaContent: Array<mediaContent>;
        public ProfileTypePermissions: Array<profileType>;
        public NotAuthorizedUsers: Array<user>;
        public FolderMediaContentMetaDataField: Array<folderMediaContentMetaDataField>;
        public ContentTypeDefinitionFolder: Array<contentTypeDefinitionFolder>;
        public ContentTypeDefinitions: Array<contentTypeDefinition<contentTypeDefinitionField>>;
        public ContentTypeDefinitionId: number;
        public Templates: Array<template>;
        public Inherit: boolean;
        public IsNew: boolean;
        public ParentArray: Array<folder<T>>;
        public ChildrenTotalCount: number;
        public ContentsTotalCount: number;
        public MediaContentTotalCount: number;
        public IsHidden: boolean;

        constructor(obj?: folder<T>) {
            super(obj);
            this.ParentId = 0;
            this.Name = '';
            this.Description = '';
            this.Parent = null;
            this.Children = new Array<folder<T>>();
            this.Contents = new Array<content>();
            this.FolderPath = '';
            this.MetaDataFields = new Array<folderMetaDataField>();
            this.MediaContent = new Array<mediaContent>();
            this.ProfileTypePermissions = new Array<profileType>();
            this.NotAuthorizedUsers = new Array<user>();
            this.FolderMediaContentMetaDataField = new Array<folderMediaContentMetaDataField>();
            this.ContentTypeDefinitionFolder = new Array<contentTypeDefinitionFolder>();
            this.ContentTypeDefinitions = new Array<contentTypeDefinition<contentTypeDefinitionField>>();
            this.ContentTypeDefinitionId = 0;
            this.Templates = new Array<template>();
            this.Inherit = true;
            this.IsNew = true;
            this.ParentArray = new Array<folder<T>>();
            this.ChildrenTotalCount = 0;
            this.ContentsTotalCount = 0;
            this.MediaContentTotalCount = 0;
            this.IsHidden = false;
            if (obj !== undefined && obj != null) {
                this.construct(obj);
            }
        }

        public construct(data: any): void {
            super.construct(data);
            this.ParentId = this.getValue<number>(data, "ParentId", 0);
            this.Name = this.getValue<string>(data, "Name", '');
            this.Description = this.getValue<string>(data, "Description", '');
            this.Parent = this.getConstructEntityValue<folder<T>>(data, "Parent", new folder<T>());
            this.Children = this.getArrayConstructEntityValue<folder<T>>(data, "Children", new Array<folder<T>>(), new folder<T>());
            this.Contents = this.getArrayConstructEntityValue<content>(data, "Contents", new Array<content>(), new content());
            this.FolderPath = this.getValue<string>(data, "FolderPath", '');
            this.MetaDataFields = this.getArrayConstructEntityValue<folderMetaDataField>(data, "MetaDataFields", new Array<folderMetaDataField>(), new folderMetaDataField());
            this.MediaContent = this.getArrayConstructEntityValue<mediaContent>(data, "MediaContent", new Array<mediaContent>(), new mediaContent());
            this.ProfileTypePermissions = this.getArrayConstructEntityValue<profileType>(data, "ProfileTypePermissions", new Array<profileType>(), new profileType());
            this.NotAuthorizedUsers = this.getArrayConstructEntityValue<user>(data, "NotAuthorizedUsers", new Array<user>(), new user());
            this.FolderMediaContentMetaDataField = this.getArrayConstructEntityValue<folderMediaContentMetaDataField>(data, "FolderMediaContentMetaDataField", new Array<folderMediaContentMetaDataField>(), new folderMediaContentMetaDataField());
            this.ContentTypeDefinitionFolder = this.getArrayConstructEntityValue<contentTypeDefinitionFolder>(data, "ContentTypeDefinitionFolder", new Array<contentTypeDefinitionFolder>(), new contentTypeDefinitionFolder());
            this.ContentTypeDefinitions = this.getArrayConstructEntityValue<contentTypeDefinition<contentTypeDefinitionField>>(data, "ContentTypeDefinitions", new Array<contentTypeDefinition<contentTypeDefinitionField>>(), new contentTypeDefinition<contentTypeDefinitionField>());
            this.ContentTypeDefinitionId = this.getValue<number>(data, "ContentTypeDefinitionId", 0);
            this.Templates = this.getArrayConstructEntityValue<template>(data, "Templates", new Array<template>(), new template());
            this.Inherit = this.getValue<boolean>(data, "Inherit", true);
            this.IsNew = this.getValue<boolean>(data, "IsNew", true);
            this.ParentArray = helpers.loadParentArray(this, 'Name', 'FolderPath')
            this.ChildrenTotalCount = this.getValue<number>(data, "ChildrenTotalCount", 0);
            this.ContentsTotalCount = this.getValue<number>(data, "ContentsTotalCount", 0);
            this.MediaContentTotalCount = this.getValue<number>(data, "MediaContentTotalCount", 0);
            this.IsHidden = this.getValue<boolean>(data, "IsHidden", false);
        }

        public clone(): folder<T> {
            return new folder<T>(this);
        }
    }
}

File: MD.CMS.BusinessLogic.Core/DataAccess/Entities/Folder.cs
using MD.CMS.BusinessLogic.Core.DataAccess.Entities.Base;
using MD.CMS.BusinessLogic.Core.Helpers.Collections;
using System.Collections.Generic;
using MD.Tools.BaseDataAccess.Plugins.Core.Mapping.ReportDefinition.ReportDesigner.CustomAttributes;
using System;
using MD.Tools.Helpers.Core.Logging;

namespace MD.CMS.BusinessLogic.Core.DataAccess.Entities
{

    public class Folder<T> : BaseHierarchycEntity<Folder<T>, long>, IHierarchyData
        where T : Content
    {
        #region Attributes
        private long _parentId;
        private string _name;
        private string _description;
        private Folder<T> _parent;
        private string _folderPath;
        private bool _inherit;
        private int _childrenTotalCount;
        private int _contentsTotalCount;
        private int _mediaContentTotalCount;

        private List<Folder<T>> _children;
        private List<T> _contents;
        private List<ContentTypeDefinition<ContentTypeDefinitionField>> _contentTypeDefinitions;
        private List<FolderMetaDataField> _metaDataFields;
        private List<MediaContent> _mediacontent;
        private List<FolderMediaContentMetaDataField> _mediaContentmetaDataFields;
        private List<ContentTypeDefinitionFolder> _contentTypeDefinitionFolder;
        private List<ProfileType> _profilePermissions;
        private List<User> _notAuthorizedUsers;
        private long _contentTypeDefinitionsId;
		private List<Template> _templates;
		private List<ContentTypeDefinitionFolderDataBoundCondition> _contentTypeDefinitionFolderDataBoundCondition;
		
		#endregion

		#region Properties

		[OmitPropertyFromReport]
        public List<ProfileType> ProfilePermissions
        {
            get
            {
                if (_profilePermissions == null) _profilePermissions = new List<ProfileType>();
                return _profilePermissions;
            }
            set { _profilePermissions = value; }
        }

        [OmitPropertyFromReport]
        public List<User> NotAuthorizedUsers
        {
            get
            {
                if (_notAuthorizedUsers == null) _notAuthorizedUsers = new List<User>();
                return _notAuthorizedUsers;
            }
            set { _notAuthorizedUsers = value; }
        }

        [OmitPropertyFromReport]
        public List<T> Contents
        {
            get
            {
                if (_contents == null)
                {
                    _contents = new List<T>();
                }
                return _contents;
            }
            set { _contents = value; }
        }

        [OmitPropertyFromReport]
        public List<ContentTypeDefinitionFolder> ContentTypeDefinitionFolder
        {
            get
            {
                if (_contentTypeDefinitionFolder == null)
                {
                    _contentTypeDefinitionFolder = new List<ContentTypeDefinitionFolder>();
                }
                return _contentTypeDefinitionFolder;
            }
            set { _contentTypeDefinitionFolder = value; }
		}

		[OmitPropertyFromReport]
		public List<ContentTypeDefinition<ContentTypeDefinitionField>> ContentTypeDefinitions
		{
			get
			{
				if (_contentTypeDefinitions == null)
				{
					_contentTypeDefinitions = new List<ContentTypeDefinition<ContentTypeDefinitionField>>();
				}
				return _contentTypeDefinitions;
			}
			set { _contentTypeDefinitions = value; }
		}

		[OmitPropertyFromReport]
		public List<ContentTypeDefinitionFolderDataBoundCondition> ContentTypeDefinitionFolderDataBoundCondition { get => _contentTypeDefinitionFolderDataBoundCondition; set => _contentTypeDefinitionFolderDataBoundCondition = value; }


		[OmitPropertyFromReport]
        public List<MediaContent> MediaContent
        {
            get
            {
                if (_mediacontent == null)
                {
                    _mediacontent = new List<MediaContent>();
                }
                return _mediacontent;
            }
            set { _mediacontent = value; }
        }

        [OmitPropertyFromReport]
        public List<FolderMetaDataField> MetaDataFields
        {
            get
            {
                if (_metaDataFields == null)
                {
                    _metaDataFields = new List<FolderMetaDataField>();
                }
                return _metaDataFields;
            }
            set { _metaDataFields = value; }
        }

        [OmitPropertyFromReport]
        public List<FolderMediaContentMetaDataField> FolderMediaContentMetaDataField
        {
            get
            {
                if (_mediaContentmetaDataFields == null)
                {
                    _mediaContentmetaDataFields = new List<FolderMediaContentMetaDataField>();
                }
                return _mediaContentmetaDataFields;
            }
            set { _mediaContentmetaDataFields = value; }
        }

        [OmitPropertyFromReport]
        public List<Template> Templates
        {
            get
            {
                if (_templates == null)
                {
                    _templates = new List<Template>();
                }
                return _templates;
            }
            set { _templates = value; }
        }

        public long ContentTypeDefinitionsId
        {
            get { return _contentTypeDefinitionsId; }
            set { _contentTypeDefinitionsId = value; }
        }

        public long ParentId
        {
            get { return _parentId; }
            set { _parentId = value; }
        }
        public bool Inherit
        {
            get { return _inherit; }
            set { _inherit = value; }
        }

        public string Name
        {
            get { return _name; }
            set { _name = value; }
        }

        public string Description
        {
            get { return _description; }
            set { _description = value; }
        }

        public override Folder<T> Parent
        {
            get { return _parent; }
            set { _parent = value; }
        }

        public string FolderPath
        {
            get { return _folderPath; }
            set { _folderPath = value; }
        }

        [OmitPropertyFromReport]
        public override List<Folder<T>> Children
        {
            get
            {
                if (_children == null)
                {
                    _children = new List<Folder<T>>();
                }
                return _children;
            }
            set { _children = value; }
        }

        [OmitPropertyFromReport]
        public bool IsNew
        {
            get
            {
                return Id.Equals(default(long));
            }
        }

        public int ChildrenTotalCount
        {
            get
            {
                return _childrenTotalCount;
            }
            set
            {
                _childrenTotalCount = value;
            }
        }

        public int ContentsTotalCount
        {
            get
            {
                return _contentsTotalCount;
            }
            set
            {
                _contentsTotalCount = value;
            }
        }

        public int MediaContentTotalCount
        {
            get { return _mediaContentTotalCount; }
            set { _mediaContentTotalCount = value; }
        }

        public bool IsHidden
        {
            get
            {
                try
                {
                    return !string.IsNullOrEmpty(Name) && Name.StartsWith('.');
                } 
                catch(Exception e)
                {
                    typeof(Folder<T>).Log(e);
                }
                return false;
            }
        }

		public override string ToString()
        {
            return Name;
        }
      
        #endregion
    }
}
